native/win_install.ps1

```diff
@@ -1,5 +1,6 @@
 Param (
     [switch]$Uninstall = $false,
+    [switch]$UsePython= $false,
```

Comment:
Aren't .exe files more likely to trigger antivirus / corporate "don't trust foreign executables" policies than the Python script?

In which case, I think it would make more sense to default to using Python unless `py -3` isn't on the path, in which case we should fall back to the executable.

What do you think?

Reply:
Hi @bovine3dom 

> Aren't .exe files more likely to trigger antivirus / corporate "don't trust foreign executables" policies than the Python script?

IMHO, this is more of a "perception" issue, and not completely unbiased. We have pre-compiled binaries even on popular [Linux][1] and [BSD][2] distributions these days. So, "compiled-binary" I think is becoming a norm even outside of Windows now.

Besides, the "trust-chain" has to bootstrap somewhere, and very few people actually download and compile their own [CPython][3] or the OS kernel. So, essentially the argument becomes somewhat like "we don't trust _'your'_ EXE (_i.e._ native_main.exe), but we trust _'their'_ EXE (_i.e._ python.exe)".

AFAIK, the best way to break this "logjam" is via [code signing][7]. I have just checked in binary signing and verification support with OpenPGP [here][4]. Please have a look. 

Windows has their own signing implementation via ["sigtool"][8]. But I think [OpenPGP][9] is more trustworthy, open-source and cross-platform friendly.

-------------------------------------------------------------------

> In which case, I think it would make more sense to default to using Python unless py -3 isn't on the path, in which case we should fall back to the executable.

My rationale for preferring EXE over PY on Windows was that we'll not only get to remove the Python dependency, we'll also get to avoid getting blames for Tridactyl problems when Python installation on Windows changes, breaks _etc._

Please note that Python on Windows is _not_ a very common set-up. I mean that it might be common within the Tridactyl community, considering that we agree that Tridactyl/Vim users are slightly more technical than the average, and are more likely to have Python on their systems. But given prior observation my guess is that a vast majority of Windows users are not likely to have Python. So I think it really boils down to who we define to be our target user-base.

With that said, I'm okay to go with the following strategy:

  - As default, look for `py -3` when invoking `win_install.ps1` on Windows

      - If Python is detected, we'd install `native_main.py`
      - if Python is _not_ detected, we'd install `native_main.exe`

As said, the risk here is that we remain dependent on the Python installation. Anytime, Python breaks, Tridactyl would also break.

Please confirm. I can submit another patch to realise the above strategy. Thanks.

-------------------------------------------------------------------

_P.S._

I did some _very quick_ research on extensions on Firefox/Chrome using native-messenger, and I find that most of them guide users to install a binary messenger, even on Mac. Here's a list of bunch of extensions using [binary native messenger][5]. They seem to be using [this][6] native-client written in Node.

Hope this helps. Thanks.

[1]: https://www.debian.org/distrib/packages
[2]: https://www.openbsd.org/faq/ports/ports.html
[3]: https://github.com/python/cpython
[4]: https://github.com/gsbabil/tridactyl/commit/2f8689cf3d78c85e104a243801813c8231751b09
[5]: https://add0n.com/open-in.html
[6]: https://github.com/andy-portmen/native-client
[7]: https://en.wikipedia.org/wiki/Code_signing
[8]: https://msdn.microsoft.com/en-us/library/windows/desktop/aa387764(v=vs.85).aspx
[9]: https://www.openpgp.org/software

Reply:
Thanks for such a comprehensive reply. I don't think we need to bother with code signing - people can just build the exe themselves and compare the checksums, which seems like fewer steps. Do you disagree?

My concern wasn't really security so much as wanting people to be able to use Tridactyl in as many places as possible. I know lots of workplaces block the running of unknown binaries (it's built into lots of antivirus software these days) so I think a python script is more likely to bypass that as, as you said, python.exe is trusted.

A related point is that we should tell users that if the native exe is blocked by their work, they should try installing python and then running our install script again.

I'm not worried about python breaking - Tridactyl will just tell people to reinstall, and that should fix it.

So, yes, please let's go with falling back to the exe.


PS: re Tridactyl users being representative of the general population - 2/3rds of them use Linux where I think the actual prevelance is 0.5%... so we're a weird bunch :)

Reply:
Hi @bovine3dom,

Apologies. This is another long reply. Please take a deep breath and see below. :-)

#### ## Prefer Python over Compiled EXE

> - As default, look for py -3 when invoking win_install.ps1 on
>   Windows
>
>     - If Python is detected, we'd install native_main.py
>     - if Python is not detected, we'd install native_main.exe

I just committed the default strategy discussed above [in this commit][0].  Please have a look.

[0]: https://github.com/gsbabil/tridactyl/commit/a940453958babe2932f8e07b9b355aab6b0c8458

-------------------------------------------------------------------

#### ## EXE Distribution and Code-Signing

> I don't think we need to bother with code signing - people can just build the exe themselves and compare the checksums, which seems like fewer steps. Do you disagree?

I'm sorry, but I'd have to disagree with this given the following. Please see details below. 

1. [Circular Dependency][2]

    - [PyInstaller][1] is a Python module, and usually installed via `python3 -m pip install --upgrade pyinstaller`

    - So, if the user wanted to compile and then verify the `native_main.exe`, they would need a functional Python 3 environment first. And if the user had Python 3 installed already (ready for PyInstaller compilation), they wouldn't need the `native_main.exe`.

2. **Non**-[Reproducible Build][3]

    - PyInstaller compiled binaries seem to be non-reproducible, meaning that if we compiled the exact same `native_main.py` at two different points in time even on the same machine, the EXE will have the same effective functionalities as the unchanged `native_main.py`, but will have slight variations in the output EXE, leading to drastically different SHA-256 digest.

    - The "non-reproducible" behavior can be verified on Windows by running the following command:

        ```
         PYTHONHASHSEED=1 pyinstaller \
          --clean \
          --console \
          --onefile \
          --noupx \
          --noconfirm \
          --workpath native \
          --distpath native 'native/native_main.py' 2> /dev/null \
          && sha256sum.exe native/native_main.exe
        ```

    - Please note that I have tried using the same `PYTHONHASHSEED=1` value advised in the [PyInstaller documentation][4], but it does not seem to work as described on PyInstaller version `3.3.1` _i.e._ SHA-256 hashes are still different.

In summary, if we agree to provide compiled `native_main.exe` for the Tridactyl users, I think we should _really_ adhere to code-signing. It would be the same reason AMO requires us to submit signed ZIP before official release. The GPG signature and SHA-256 sums should be released together with `native_main.exe`, and it should be up to the user to perform the verification.

[1]: https://www.pyinstaller.org
[2]: https://en.wikipedia.org/wiki/Circular_dependency
[3]: https://en.wikipedia.org/wiki/Deterministic_compilation
[4]: https://pyinstaller.readthedocs.io/en/stable/advanced-topics.html#creating-a-reproducible-build

-------------------------------------------------------------------

#### ## Windows Support and Development Directions

> Tridactyl users being representative of the general population - 2/3rds of them use Linux where I think the actual prevelance is 0.5%... so we're a weird bunch :)

I think we are getting pretty close to the hard questions below:

  1. How much support would we like to provide for Tridactyl users on Windows?
  2. Would Windows users get the same functionalities as the Linux/Mac users with Tridactyl?

Right now, I don't think Windows is the same first-class citizen that a Linux/Mac system is with Tridactyl. I think a clear decision around this will make future development directions much more coherent.

I think the active owners/maintainers of this project (I believe you and @cmcaine) should set the course for this sooner rather than later. I'm happy to listen and contribute where I can!

Hope this helps. Thank you.


_P.S._

I have submitted a bug report with PyInstaller [here][5] regarding reproducible-build. It seems that their current PyPI version `3.3.1` is buggy, but their [latest master][6] works fine. However, even if we eliminate the "reproducible-build" issue, the "circular-dependency" issue in my opinion still stands, and I'd still strongly vote in for code-signing. :-)

Thank you.

[5]: https://github.com/pyinstaller/pyinstaller/issues/3526
[6]: https://github.com/pyinstaller/pyinstaller/commit/1033a8770ab96cc8fa0f2d3047b2db3db9dcda43